ML-276 Expose C++11 Pseudo-random numbers in ECL

garonsky opened the pull request:
Signed-off-by: Gleb Aronsky gleb.aronsky@lexisnexis.com

Commit:
ML-276 Expose C++11 Pseudo-random numbers in ECL

Signed-off-by: Gleb Aronsky <gleb.aronsky@lexisnexis.com>

garonsky commented:
@ghalliday please review
@johnholt please review

HPCCSmoketest commented:
Automated Smoketest 
Sha: 98f4879c7acbebd6f218839fe48f2a408e3842e2
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@garonsky comments follow:

This would be better implemented as part of the standard library - as external functions. rather than by extending the syntax.  Probably by creating a Std.Random which contained all the definitions.  That is the general approach for any new functionality which can be implemented without changing the language.

The syntax should be as simple to use as possible.  So for example the most important thing is the type of distribution - so that should be the first argument.  The type of random number generator is less important, and it should also be optional (using the default).

The scope of the code included in the critical block is too large.  As time goes on we are using more and more threads.  (In 6.2 I would expect to be routinely using all the cores for inline tables and projects if we can.)  So in this situation I think you need to be using a thread local variable to contain and instance of the appropriate random number generator.

(The current random() implementation probably requires revisiting.)

Other minor issues:
The constant folding needs to check that the expressions are constant before deferencing queryValue() and assuming that is non null.  Defining the functions as external functions removes that problem.

There is code that is duplicated.  E.g., the code that performs a switch on engine to return the generator.

The critical section being dependent on the engine is leaking details of the internal class implementation where it doesn't belong.  If you wanted to do this you should have a random number generator that used a single engine, and protect an array of them with an array of critical sections.

The code doesn't match the style used for member variable in the rest of the core system code base.  i.e. We do not prefix member variable with m_.

Some of the members e.g., m_pNegativeBinomialDistribution are not used.

There are some minor indentation issues.

Commit:
ML-276 Expose C++11 Pseudo-random numbers in ECL

- Changes based on comments

Signed-off-by: Gleb Aronsky <gleb.aronsky@lexisnexis.com>

garonsky commented:
I made the following changes based on the comments

-	Improved syntax to use distribution as first parameter and engine as an optional parameter
-	Removed locks and utilized TLS for new and existing generators
-	Add check for constant expression
-	Removed duplicated code by using a function template
-	Removed unused variables
-	Minor improvements to code style and indentation

@ghalliday please review.

HPCCSmoketest commented:
Automated Smoketest 
Sha: cea550887d483ff8cd765a12bba982391302a68d
Build: failed
Number of Error(s): 7
  Policy CMP0046 is not set: Error on non-existent dependency in
  Policy CMP0046 is not set: Error on non-existent dependency in
---- Error log:    /mnt/disk1/home/vamosax/smoketest/PR-8805/build/esp/src/eclwatch_build_err.txt
RPM build errors:
CPack Error: Problem copying the package: /mnt/disk1/home/vamosax/smoketest/PR-8805/build/_CPack_Packages/Linux-x86_64/RPM/hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm to /mnt/disk1/home/vamosax/smoketest/PR-8805/build/hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
CPack Error: Error when generating package: hpccsystems-platform
make: *** [package] Error 1

HPCCSmoketest commented:
Automated Smoketest 
Sha: cea550887d483ff8cd765a12bba982391302a68d
Build: failed
Number of Error(s): 7
  Policy CMP0046 is not set: Error on non-existent dependency in
  Policy CMP0046 is not set: Error on non-existent dependency in
---- Error log:    /mnt/disk1/home/vamosax/smoketest/PR-8805/build/esp/src/eclwatch_build_err.txt
RPM build errors:
CPack Error: Problem copying the package: /mnt/disk1/home/vamosax/smoketest/PR-8805/build/_CPack_Packages/Linux-x86_64/RPM/hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm to /mnt/disk1/home/vamosax/smoketest/PR-8805/build/hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
CPack Error: Error when generating package: hpccsystems-platform
make: *** [package] Error 1

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3d1526002f41dd403cf0f6e78a7ec1bffaf2fc43
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@garonsky these should be implemented as functions inside the standard library, rather than extending the language syntax.  I will comment on the code anyway - even if it would not be relevant after they had been moved.

ghalliday commented:
The pull request is missing test cases.

ghalliday commented on ecl/hql/hqlfold.cpp:
```diff
@@ -3584,8 +3584,40 @@ IHqlExpression * foldConstantOperator(IHqlExpression * expr, unsigned foldOption
+        {
+            if (foldOptions & (HFOfoldimpure|HFOforcefold))
+            {
+                if (expr->queryChild(0)->numChildren() == 0)
```
Should be expr->numChildren()

ghalliday commented on ecl/hql/hqlfold.cpp:
```diff
@@ -3584,8 +3584,40 @@ IHqlExpression * foldConstantOperator(IHqlExpression * expr, unsigned foldOption
+                if (expr->queryChild(0)->numChildren() == 0)
+                    return createConstant(expr->queryType()->castFrom(true, (__int64)rtlRandom()));
+
+                if (!expr->isConstant())
```
I suspect this will always return false.

ghalliday commented on ecl/hql/hqlfold.cpp:
```diff
@@ -3584,8 +3584,40 @@ IHqlExpression * foldConstantOperator(IHqlExpression * expr, unsigned foldOption
+                if (!expr->isConstant())
+                    break;
+
+                if (expr->queryChild(0)->queryName() == distributionUniformAtom)
```
Better to assign queryChild(0) into a local variable.  Simplifies the following code, and makes it more efficient.

(Note all this code goes if implemented as a library function)

ghalliday commented on ecl/hql/hqlgram.y:
```diff
@@ -12692,6 +12711,54 @@ featureModifiers
+    ;
+
+pseudo_random_number_distribution
+    : UNIFORM ',' expression ',' expression
```
UNIFORM(a,b) would be the more consistent syntax rather than UNIFORM,a,b

ghalliday commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -3124,8 +3124,28 @@ void HqlCppTranslator::buildExpr(BuildCtx & ctx, IHqlExpression * expr, CHqlBoun
     case no_random:
-        doBuildExprSysFunc(ctx, expr, tgt, rtlRandomId);
+    {
+        if (expr->queryChild(0)->numChildren() == 0)
```
again, save queryChild(0) and distribution->queryName() into temporaries.

ghalliday commented on rtl/eclrtl/eclrtl.cpp:
```diff
@@ -3762,6 +3766,45 @@ void rtlSeedRandom(unsigned value)
+unsigned rtlPseudoRandomNumberBinomialDistribution(unsigned engine, double probablity, unsigned upper_bound)
+{
+    if (engine >= IPseudoRandomNumberGenerator::ePseudoRandomNumberEngine::NUMBER_OF_ENGINES)
+        return rtlRandom(); // Unknown engine, so just use default
```
Surely picking a default engine would be better - otherwise it will not match the distribution.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -812,27 +814,157 @@ class CRandom: public CInterface, public IRandomNumberGenerator
+};
 
-static CriticalSection gobalRandomSect;
+thread_local Owned<IRandomNumberGenerator> RandomMain = createRandomNumberGenerator();
```
I suspect this doesn't compile on osx.

ghalliday commented on system/jlib/jutil.cpp:
```diff
@@ -812,27 +814,157 @@ class CRandom: public CInterface, public IRandomNumberGenerator
 {
-    CriticalBlock block(gobalRandomSect); // this is a shame but it is not thread-safe without
-    return RandomMain.next();
+    return RandomMain->next();
```
This is changing the implementation of semantics of an existing function.  It would need careful checking - e.g., how are they seeded?

ghalliday commented on rtl/eclrtl/eclrtl.cpp:
```diff
@@ -48,15 +48,19 @@
 IRandomNumberGenerator * random_;
 static CriticalSection random_Sect;
 
+thread_local Owned<IPseudoRandomNumberGenerator> pseudo_random_ = createPseudoRandomNumberGenerator();
```
I'm not sure if this will compile on osx.

garonsky commented:
Will update and re-open based on comments.
